Fix typing errors in torch.distributed.distributed_c10d.*

xuzhao9 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47534 Fix typing errors in torch.distributed.*, close issue #42967.
* #47533 Fix typing errors in torch.distributed.nn.* directory.
* **#47532 Fix typing errors in torch.distributed.distributed_c10d.***
* #47531 Fix type annotation errors in torch.distributed.* directory

Differential Revision: [D24952501](https://our.internmc.facebook.com/intern/diff/D24952501)

xuzhao9 committed:
Fix typing errors in torch.distributed.distributed_c10d.*

[ghstack-poisoned]

xuzhao9 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -117,11 +120,11 @@ def register_backend(cls, name, func):
 
 
-class reduce_op(object):
+class _reduce_op(object):
```
Need to rename this class because reduce_op is later assigned to an instance of this class.

xuzhao9 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1335,7 +1360,7 @@ def all_gather_multigpu(output_tensor_lists,
 def _object_to_tensor(obj):
     buffer = pickle.dumps(obj)
-    byte_storage = torch.ByteStorage.from_buffer(buffer)
+    byte_storage = torch.ByteStorage.from_buffer(buffer)  # type: ignore[attr-defined]
```
Neither mypy nor I can find from_buffer() function in torch.ByteStorage.

xuzhao9 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1389,7 +1414,7 @@ def all_gather_object(object_list, obj, group=group.WORLD):
     group_backend = get_backend(group)
     is_nccl_backend = group_backend == Backend.NCCL
-    current_device = torch.device("cpu")
+    current_device: Union[int, torch.device] = torch.device("cpu")
```
torch.cuda.current_device() returns int, and torch.device() returns torch.device.

malfet replied:
Then we should fix how torch_device is redefined on line 1422 (i.e. assign it as `current_device = torch.device('cuda', torch.cuda.current_device())`
```suggestion
    current_device = torch.device("cpu")
```

xuzhao9 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1400,7 +1425,7 @@ def all_gather_object(object_list, obj, group=group.WORLD):
     # until the correct size when deserializing the tensors.
     group_size = get_world_size(group=group)
-    object_sizes_tensor = torch.zeros(group_size, dtype=int, device=current_device)
+    object_sizes_tensor = torch.zeros(group_size, dtype=int, device=current_device)  # type: ignore[call-overload]
```
If is_nccl_backend is set, current_device is an int, which will not be accepted by torch.zeros().

xuzhao9 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1410,7 +1435,7 @@ def all_gather_object(object_list, obj, group=group.WORLD):
     input_tensor.resize_(max_object_size)
     coalesced_output_tensor = torch.empty(
-        max_object_size * group_size, dtype=torch.uint8, device=current_device
+        max_object_size * group_size, dtype=torch.uint8, device=current_device  # type: ignore[arg-type]
```
If is_nccl_backend is set, current_device is an int, which will not be accepted by torch.empty().

xuzhao9 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1420,7 +1445,7 @@ def all_gather_object(object_list, obj, group=group.WORLD):
     # Deserialize outputs back to object.
     for i, tensor in enumerate(output_tensors):
-        tensor = tensor.type(torch.ByteTensor)
+        tensor = tensor.type(torch.ByteTensor)  # type:ignore[call-overload]
```
tensor.type() accepts dtype, but torch.ByteTensor is typing.Type.

xuzhao9 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1559,14 +1584,14 @@ def broadcast_object_list(object_list, src, group=group.WORLD):
         current_device = torch.cuda.current_device()
-        object_sizes_tensor = object_sizes_tensor.to(current_device)
-        object_sizes_tensor = object_sizes_tensor.to(current_device)
+        object_sizes_tensor = object_sizes_tensor.to(current_device)  # type: ignore[call-overload]
```
If is_nccl_backend is set, current_device is an int, which will not be accepted by torch.Tensor.to().

malfet commented on torch/_C/_distributed_c10d.pyi:
```diff
@@ -122,7 +122,7 @@ class Work:
     def is_success(self) -> bool: ...
     def exception(self) -> Any: ...
-    def wait(self, timeout: timedelta) -> bool: ...
+    def wait(self, timeout: timedelta = ...) -> bool: ...
```
Didn't know ellipsis could be used for default values, but wouldn't it be better to give it an actual default?

xuzhao9 replied:
> Didn't know ellipsis could be used for default values, but wouldn't it be better to give it an actual default?

The default value (kNoTimeout in distributed/c10d/init.cpp) is not exported to Python, should I export this value and add it as the default?

walterddr replied:
make senes to add it since default no timeout is actually affecting lots of internal behavior of processgroup

malfet commented on torch/distributed/distributed_c10d.py:
```diff
@@ -226,6 +229,7 @@ def _get_group_size(group):
     """
     if group is GroupMember.WORLD:
         _check_default_pg()
+        assert _default_pg is not None
```
Wouldn't it be a bit cleaner, if _check_default_pg() would return an non-optional PG which one then binds to a local variable?

walterddr replied:
similar to 
```
def _check_default_pg():
  if _default_pg is not None:
    return _default_pg
  else: 
    raise RuntimeError("Default process group is not initialized")
```

xuzhao9 replied:
Good point, fixed by returning ProcessGroup in _check_default_pg(), and assign the return value to a local variable "default_pg".

malfet commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1559,14 +1584,14 @@ def broadcast_object_list(object_list, src, group=group.WORLD):
         # See note about using torch.cuda.current_device() here in docstring.
         # We cannot simply use my_rank since rank == device is not necessarily
         # true.
         current_device = torch.cuda.current_device()
```
This would be a correct fix here
```suggestion
        current_device = torch.device('cuda', torch.cuda.current_device())
```

walterddr commented on torch/distributed/distributed_c10d.py:
```diff
@@ -341,12 +346,15 @@ def get_backend(group=group.WORLD):
     _check_default_pg()
 
     if group == GroupMember.WORLD:
+        assert _default_pg is not None
         pg = _default_pg
```
we can replaced all of these with local variable
```suggestion
        pg = _check_default_pg()
```
if we adopt the above method, and simplify those changes below

walterddr commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1465,7 +1490,7 @@ def gather_object(obj, object_gather_list=None, dst=0, group=group.WORLD):
     input_tensor, local_size = _object_to_tensor(obj)
     group_backend = get_backend(group)
-    current_device = torch.device("cpu")
+    current_device: Union[int, torch.device] = torch.device("cpu")
```
same here

walterddr commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1474,7 +1499,7 @@ def gather_object(obj, object_gather_list=None, dst=0, group=group.WORLD):
     # until the correct size when deserializing the tensors.
     group_size = get_world_size(group=group)
-    object_sizes_tensor = torch.zeros(group_size, dtype=int, device=current_device)
+    object_sizes_tensor = torch.zeros(group_size, dtype=int, device=current_device)  # type: ignore
```
nit: missing ignore type

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.distributed_c10d.*"




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.distributed_c10d.*"




[ghstack-poisoned]

walterddr commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1212,6 +1212,7 @@ that adds a prefix to each key inserted to the store.
       py::call_guard<py::gil_scoped_release>());
 
   module.attr("_DEFAULT_FIRST_BUCKET_BYTES") = ::c10d::kDefaultFirstBucketBytes;
+  module.attr("_DEFAULT_NO_TIMEOUT") = py::cast(kNoTimeout);
```
nit: was wondering what's the value add here - this means we need to separately maintain the defaults on cpp and python side anyway (e.g. ppl who changed the signature on cpp side from `kNoTimeout` to `kDefaultTimeout` for example, doesn't automatically updates the python signature. 

if this is the case it might be just easier to put `_DEFAULT_NO_TIMEOUT: timedelta = 0` on the pyi file?

xuzhao9 replied:
I don't think we can do that, since everything defined in the pyi file needs to be explicitly exported from the cpp file. Otherwise there will be a test error.

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.distributed_c10d.*"




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.distributed_c10d.*"




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.distributed_c10d.*"


Differential Revision: [D24952501](https://our.internmc.facebook.com/intern/diff/D24952501)

[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.distributed_c10d.*"


Differential Revision: [D24952501](https://our.internmc.facebook.com/intern/diff/D24952501)

[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.distributed_c10d.*"


Differential Revision: [D24952501](https://our.internmc.facebook.com/intern/diff/D24952501)

[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.distributed_c10d.*"


Differential Revision: [D24952501](https://our.internmc.facebook.com/intern/diff/D24952501)

[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.distributed_c10d.*"


Differential Revision: [D24952501](https://our.internmc.facebook.com/intern/diff/D24952501)

[ghstack-poisoned]
